include/boost/windows_sspi/detail/sspi_impl.hpp

```diff
@@ -119,11 +135,28 @@ class sspi_decrypt {
         encrypted_data = std::vector<char>(reinterpret_cast<const char*>(pExtraBuffer->pvBuffer), reinterpret_cast<const char*>(pExtraBuffer->pvBuffer) + pExtraBuffer->cbBuffer);
       }
     }
-    return sc;
+    BOOST_ASSERT(!decrypted_data.empty());
+
+    return state::data_available;
+  }
+
+  // TODO: Consider making this more flexible by not requering a
+  // vector of chars, but any view of a range of bytes
+  void put(const std::vector<char>& data) {
+    encrypted_data.insert(encrypted_data.end(), data.begin(), data.end());
+  }
+
+  std::vector<char> get(std::size_t max) {
```

Comment:
returning a new vector here forces memory allocation. I wonder whether there is room for another version of this function that takes a reference to a vector to modify? This way, the caller is able to avoid the pessimisation by re-using the same vector on each call.